PP-773: Array jobs do not send emails when specified for -m abe

vchlum opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
* No matter the -m abe mail option is supplied, array jobs (both parent jobs and subjobs) do not send e-mails.
* [PP-773](https://pbspro.atlassian.net/browse/PP-773)

#### Affected Platform(s) and PBS Version
* Platform: all
* Bugs: 18.1.0

#### Cause / Analysis / Design
* Please, see the [project documentation area](https://pbspro.atlassian.net/wiki/spaces/PD/pages/172884019/PP-773+Array+jobs+do+not+send+emails+when+specified+for+-m+abe)
* Please, see the [pbspro community forum](http://community.pbspro.org/t/pp-773-array-jobs-do-not-send-emails-when-specified-for-m-abe/844)

#### Solution Description
* Sending e-mails by subjobs was enabled by putting JOB_ATR_mailpnts into attrs_to_copy[]. The svr_mailowner_id() will send the emails only if the 'j' mailpoint is supplied.
* E-mails from parent job are sent using svr_mailowner(). The function is called from the same places as the accounting logs ('S' and 'E') are created.
* Test log: [PP-773_ptl_output.txt](https://github.com/PBSPro/pbspro/files/1678438/PP-773_ptl_output.txt)


#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

mike0042 commented on src/cmds/qsub.c:
```diff
@@ -3723,6 +3723,13 @@ process_opts(int argc, char **argv, int passet)
 #endif
+
+	if (errflg == 0 && J_opt == 0 && get_attr(attrib, ATTR_m, NULL) != NULL &&
+		strchr(get_attr(attrib, ATTR_m, NULL), 'j') != (char *)0) {
```
Please compare the result of strchr() against NULL rather than (char *)0.

vchlum replied:
Done

mike0042 commented on src/lib/Libecl/ecl_verify_values.c:
```diff
@@ -741,6 +741,9 @@ verify_value_mailpoints(int batch_request, int parent_object, int cmd,
+
 	if (strcmp(pattr->value, "n") != 0) {
 		pc = pattr->value;
 		while (*pc) {
```
While you're making changes to this area, the loop is better suited as "for" than "while".

vchlum replied:
Done

mike0042 commented on src/server/req_runjob.c:
```diff
@@ -1139,6 +1139,8 @@ complete_running(job *jobp)
 			if (parent->ji_wattr[(int)JOB_ATR_depend].at_flags&ATR_VFLAG_SET)
 				(void)depend_on_exec(parent);
+
+			svr_mailowner(parent, MAIL_BEGIN, MAIL_NORMAL, (char *)0);
```
Please use NULL rather than (char *)0.

vchlum replied:
Done

mike0042 commented on src/server/svr_mail.c:
```diff
@@ -606,6 +606,16 @@ svr_mailowner_id(char *jid, job *pjob, int mailpoint, int force, char *text)
+			if (pjob->ji_qs.ji_svrflags & JOB_SVFLG_SubJob) {
+				if (pjob->ji_wattr[(int)JOB_ATR_mailpnts].at_flags &ATR_VFLAG_SET) {
+					if (strchr(pjob->ji_wattr[(int)JOB_ATR_mailpnts].at_val.at_str,
+						MAIL_SUBJOB) == (char *)0)
```
Please use NULL rather than (char *)0.

vchlum replied:
Done

mike0042 commented on test/tests/functional/pbs_array_job_mail.py:
```diff
@@ -0,0 +1,85 @@
+                    emailpass = 1
+
+            self.assertTrue(emailpass, "Message '" + jobid + " " + msg +
+                            "' not found in " + mailfile)
```
Please add one more test with ATTR_m set to "j" and confirm you receive an error message.

vchlum replied:
New test added

vchlum commented:
[PP-773_ptl_output2.txt](https://github.com/PBSPro/pbspro/files/1680977/PP-773_ptl_output2.txt)

vchlum commented:
Please, could I get one more review? This PR needs one more...

vchlum commented on test/tests/functional/pbs_array_job_mail.py:
```diff
@@ -0,0 +1,106 @@
+                           parent_jid, max_attempts=15, interval=2)
+
+        self.logger.info("Wait 10s for array job to finish")
+        time.sleep(10)
```
OK, I use expect now, but we also need to wait for saving the emails into the mail file, which takes some time. I did some tests and 5s is not enough on my testbed. So, the time.sleep(10) is necessary anyway.

vchlum commented:
@bhroam Thank you very much for your review. I have addressed your comments.
[PP-773_ptl_output3.txt](https://github.com/PBSPro/pbspro/files/1735887/PP-773_ptl_output3.txt)

mike0042 commented:
I sign off on the latest changes.

vchlum commented:
The commits are squashed and branch rebased. It is ready for merge @PBSPro/pbspro-maintainers .

mike0042 commented:
Merged. Thank you for your contribution!
